### PR TITLE
Ec2 aws customer gateway

### PR DESCRIPTION
**Pull request by MichaelBaydoun**
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

ec2_customer_gateway
##### Summary:

Couldn't find a module to create a customer gateway so wrote one

Take is easy on me.  First module contribution, first python code, first boto use.

---

This change is [Reviewable](https://reviewable.io/reviews/ansible/ansible-modules-extras/1868)

### CODE REVIEW

**Comment by resmo** [comment]
Thanks @MichaelBaydoun for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

**Comment by wimnat** [comment]
New aws modules should be using boto3.

@MichaelBaydoun can you rewrite to use boto3?

I'll try and poke some people to get the PR merged which has the doc update on this.

**Comment by MichaelBaydoun** [comment]
Sure, I will rewrite using boto3

**Comment by MichaelBaydoun** [comment]
@wimnat module has been rewritten using boto3

**Comment by MichaelBaydoun** [comment]
Discovered this module duplicates functionality of another new module submission https://github.com/ansible/ansible-modules-extras/pull/1337

**Comment by gundalow** [comment]
1) Few bits of trailing whitespace
The following command will fix that up:

`sed -i 's/[ \t]*$//' ec2_customer_gateway.py`

2) pep8 -r --ignore=E501,E221,W291,W391,E302,E251,E203,W293,E231,E303,E201,E225,E261,E241,E402 ec2_customer_gateway.py
`ec2_customer_gateway.py:114:65: E226 missing whitespace around arithmetic operator`

**Comment by MichaelBaydoun** [comment]
Based on feedback during the new module meeting today, and rereading the Amazon module guidelines, I'm tempted to rename this module vpc_customer_gateway.  Thoughts?

**Comment by thaumos** [comment]
@MichaelBaydoun, we have other modules called ec2_vpc_*...  Should we keep to the naming convention as the rest of the modules?

**Comment by MichaelBaydoun** [comment]
@thaumos great question, I started with ec2_ as you can see. 

The amazon module guidelines [https://github.com/ansible/ansible-modules-extras/blob/devel/cloud/amazon/GUIDELINES.md](url) state "Base the name of the module on the part of AWS that you actually use. (A good rule of thumb is to take whatever module you use with boto as a starting point)."  

In boto, I'm using ec2.create_customer_gateway for example, so you're right, by that guideline it should start with ec2_.  

The guidelines also state "Don't further abbreviate names - if something is a well known abbreviation due to it being a major component of AWS, that's fine, but don't create new ones independently (e.g. VPC, ELB, etc. are fine)."  

By that measure, I should probably stick with ec2_customer_gateway, unless anyone feels strongly another way.

File: cloud/amazon/ec2_customer_gateway.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+short_description: Manage an AWS customer gateway
+description:
+    - Manage an AWS customer gateway
+version_added: "2.1"
+author: Michael Baydoun (@MichaelBaydoun)
+requirements: [ boto3 ]
+options:
+  bgp_asn:
+    description:
+      - Border Gateway Protocol (BGP) Autonomous System Number (ASN), required when state=present
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: `when C(state=present).`

Note full stop at the end.

File: cloud/amazon/ec2_customer_gateway.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    required: false
+    default: null
+  ip_address:
+    description:
+      - Internet-routable IP address for customers gateway, must be a static address
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: Please add a full stop

File: cloud/amazon/ec2_customer_gateway.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    required: true
+  name:
+    description:
+      - Name of the customer gateway
+    required: true 
+  state:
+    description:
+      - Create or terminate the Customer Gateway
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: Please add a full stop

File: cloud/amazon/ec2_customer_gateway.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+    def __init__(self, module):
+        self.module = module
+
+        try:
+            region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+            if not region:
+                module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+            self.ec2 = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        except boto.exception.NoAuthHandlerFound, e:
```
Comment (bob-smith) [inline cloud/amazon/ec2_customer_gateway.py]: I think this should be `boto3`

```
  File "/blahblahablah/ec2_customer_gateway", line 114, in __init__
    except boto.exception.NoAuthHandlerFound, e:
NameError: global name 'boto' is not defined

```

File: cloud/amazon/ec2_customer_gateway.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+      - Create or terminate the Customer Gateway
+    required: false
+    default: present
+    choices: [ 'present', 'absent' ]
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py:52]: Examples need work
- No `name` (required parameter)

File: cloud/amazon/ec2_customer_gateway.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+DOCUMENTATION = '''
+---
+module: ec2_customer_gateway
+short_description: Manage an AWS customer gateway
+description:
+    - Manage an AWS customer gateway
+version_added: "2.1"
+author: Michael Baydoun (@MichaelBaydoun)
+requirements: [ boto3 ]
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: I think you need `botocore` for exception handling as well
`requirements: [ botocore, boto3 ]`

File: cloud/amazon/ec2_customer_gateway.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            returned: when gateway exists and is available 
+            sample: ipsec.1 
+            type: string
+'''
+
+try:
+    import boto3
+    HAS_BOTO3 = True
+except ImportError:
+    HAS_BOTO3 = False
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py:119]: Half way there, the module will currently fail with a backtrace if boto3 isn't installed. This isn't clear in the Developing Module guide. I'll get that updated.

I think you actually need `boto` and `boto3`

```
try:
    import botocore
    HAS_BOTOCORE = True
except ImportError:
    HAS_BOTOCORE = False

try:
    import boto3
    HAS_BOTO3 = True
except ImportError:
    HAS_BOTO3 = False
```

File: cloud/amazon/ec2_customer_gateway.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        dict(
+            bgp_asn = dict(required=False, type='int'),
+            ip_address = dict(required=True),
+            name = dict(required=True),
+            state = dict(default='present', choices=['present', 'absent']),
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py:197]: And here you can check if the imports were sucessful

```
    if not HAS_BOTOCORE:
      module.fail_json(msg='botocore is required.')

    if not HAS_BOTO3:
      module.fail_json(msg='boto3 is required.')

```

**Comment by gundalow** [comment]
Good script, I think this will be useful for people. It's fairly close, just a few changes needed.

If I locally fix the script to import boto and boto3 I can get it basically working.
- Confirmed that this meets http://docs.ansible.com/ansible/developing_modules.html#module-checklist
  - Confirmed this will create Customer Gateways, subject to caveats below
  - Confirmed that deletion works
  - NOT reviewed Python from a coding point of view, I'll leave that to someone else

However the checking to see if an entry already exists seems to only use the IP address. As long as the IP address is set and matches an entry in AWS the other parameters (such as Name and AS number) are not used to check.

Meaning

```
  1     - name: create first aws customer gateway
  2       ec2_customer_gateway:
  3         name: moo
  4         region: us-east-1
  5         bgp_asn: 1111
  6         ip_address: 1.1.1.1
  7       register: cgw
  8     - name: create second customer gateway
  9       ec2_customer_gateway:
 10         name: moo222222
 11         region: us-east-1
 12         bgp_asn: 222222
 13         ip_address: 1.1.1.1
 14       register: cgw

```

First task creates, second task is listed as no change. Check AWS and you only see the 1.1.1.1/1111/moo.

From looking at the AWS console interface https://console.aws.amazon.com/vpc/home?region=us-east-1#cgws: is looks like you can only change the name.

http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/ApiReference-cmd-CreateCustomerGateway.html 

states:

> Important
> You cannot create more than one customer gateway with the same VPN type, IP address, and BGP ASN parameter values. If you run an identical request more than one time, the first request creates the customer gateway, and subsequent requests return information about the existing customer gateway. The subsequent requests do not create new customer gateway resources.

This feels like an important thing, should it be added as in the `notes:`

File: cloud/amazon/ec2_customer_gateway.py (hunk 2 above)
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: "Internet-routable IPv4..."

According to https://aws.amazon.com/vpc/faqs/

**Comment by gundalow** [comment]
needs_revision

File: cloud/amazon/ec2_customer_gateway.py (hunk 4 above)
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: `botocore.exception`

**Comment by gundalow** [comment]
I've been corrected, it's `botocore` for exceptions.

**Comment by MichaelBaydoun** [comment]
New commit with all feedback fixed except the ability to change the name of an existing gateway.  I might need a touch of python help on that one.

File: cloud/amazon/ec2_customer_gateway.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    bgp_asn = module.params.get('bgp_asn')
+    ip_address = module.params.get('ip_address')
+    name = module.params.get('name')
+
+    existing = gw_mgr.describe_gateways(module.params['ip_address'])
+
+    results = dict(changed=False)
+    if module.params['state'] == 'present':
+        if existing['CustomerGateways']:
+            results['gateway']=existing
```
Comment (MichaelBaydoun) [inline cloud/amazon/ec2_customer_gateway.py:215]: Specifically, I need to evaluate the contents of existing['CustomerGateways'], find the tag named "Name" if it exists, and if it doesn't exist or it's value is not equal to the name passed in then I need to call the tag function.  Doing that evaluation is a bit beyond my current python skills.  Below is an example of the data returned.  CustomerGateways will only ever contain one item if the desired ip address already exists.

```
{
    "changed": false, 
    "gateway": {
        "CustomerGateways": [
            {
                "BgpAsn": "65000", 
                "CustomerGatewayId": "cgw-9a59b6f3", 
                "IpAddress": "1.2.3.4", 
                "State": "available", 
                "Tags": [
                    {
                        "Key": "Name", 
                        "Value": "foobar"
                    }
                ], 
                "Type": "ipsec.1"
            }
        ], 
        "ResponseMetadata": {
            "HTTPStatusCode": 200, 
            "RequestId": "7a3f781f-1c68-4169-a709-ee41ea4e8d5f"
        }
    }, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "bgp_asn": 65000, 
            "ec2_url": null, 
            "ip_address": "1.2.3.4", 
            "name": "foobar1", 
            "profile": null, 
            "region": "us-east-1", 
            "security_token": null, 
            "state": "present", 
            "validate_certs": true
        }
    }
}
```

**Comment by wimnat** [comment]
You should make use of new camel_dict_to_snake_dict function in ec2 module utils to ensure your results back from boto are changed to snake case. See aws guidelines for more info on the new helper functions

File: cloud/amazon/ec2_customer_gateway.py [hunk 10]
```diff
@@ -48,14 +48,21 @@
 '''
 
 EXAMPLES = '''
+
+# Note: You cannot create more than one customer gateway with the same IP address. If you run an identical request more than one time, the first request creates the customer gateway, and subsequent requests return information about the existing customer gateway. The subsequent requests do not create new customer gateway resources.
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: Hi,

Applogies, I wasn't clear, you can specify a note in the following form
https://github.com/ansible/ansible-modules-extras/blob/devel/system/debconf.py#L32

So it appears as below the examples, as shown here http://docs.ansible.com/ansible/debconf_module.html

File: cloud/amazon/ec2_customer_gateway.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    default: present
+    choices: [ 'present', 'absent' ]
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
+
+# Note: You cannot create more than one customer gateway with the same IP address. If you run an identical request more than one time, the first request creates the customer gateway, and subsequent requests return information about the existing customer gateway. The subsequent requests do not create new customer gateway resources.
```
Comment (gundalow) [inline cloud/amazon/ec2_customer_gateway.py]: Hi,

Applogies, I wasn't clear, you can specify a note in the following form
https://github.com/ansible/ansible-modules-extras/blob/devel/system/debconf.py#L32

So it appears as below the examples, as shown here http://docs.ansible.com/ansible/debconf_module.html

**Comment by gregdek** [comment]
Thanks @MichaelBaydoun for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by MichaelBaydoun** [comment]
Last commit adds support for changing the gateway name.  All comments above should be addressed.

**Comment by MichaelBaydoun** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @MichaelBaydoun for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gundalow** [comment]
Getting backtraces now

```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/home/johnb/.ansible/tmp/ansible-tmp-1460563602.71-185391798102327/ec2_customer_gateway", line 2440, in <module>
    main()
  File "/home/johnb/.ansible/tmp/ansible-tmp-1460563602.71-185391798102327/ec2_customer_gateway", line 217, in main
    if existing['CustomerGateways'][0]['Tags']:
KeyError: 'Tags'

fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "ec2_customer_gateway"}, "parsed": false}
```

From https://console.aws.amazon.com/vpc/home?region=us-east-1#cgws: I can see the gw has been added, but without a name

**Playbook**

```
- name: "Test ec2_customer_gateway.py"
  hosts: all
  tasks:
    - name: create first aws customer gateway
      ec2_customer_gateway:
        name: fred
        region: us-east-1
        bgp_asn: 2222
        ip_address: 2.2.2.2
        state: present
      register: cgw

```

**Comment by gundalow** [comment]
needs_revision

**Comment by MichaelBaydoun** [comment]
All bugs squashed, ready_for_review

**Comment by gundalow** [comment]
Thanks for this.

shipit

**Comment by gregdek** [comment]
@wimnat @thaumos any chance we could get a second shipit from one of you? Thanks!

**Comment by MichaelBaydoun** [comment]
Modified botocore exception handling as discussed in new module meeting today
ready_for_review

**Comment by MichaelBaydoun** [comment]
@wimnat @thaumos, would either of you have time to review before the new module meeting today?

**Comment by wimnat** [comment]
@MichaelBaydoun sorry I'm just flat out at the mo so not re-reviewed. Can you post the module output now showing the new snake case?

**Comment by wimnat** [comment]
Can you also squash your commits?

**Comment by MichaelBaydoun** [comment]
@wimnat 

```
PARSED OUTPUT
{
    "changed": true, 
    "gateway": {
        "customer_gateway": {
            "bgp_asn": "65001", 
            "customer_gateway_id": "cgw-e09d7389", 
            "ip_address": "6.7.4.9", 
            "state": "available", 
            "type": "ipsec.1"
        }, 
        "response_metadata": {
            "http_status_code": 200, 
            "request_id": "e503a70f-4ceb-4cbb-8a1b-035417ac0f25"
        }
    }, 
    "invocation": {
        "module_args": {
            "aws_access_key": null, 
            "aws_secret_key": null, 
            "bgp_asn": 65001, 
            "ec2_url": null, 
            "ip_address": "6.7.4.9", 
            "name": "miketest2", 
            "profile": null, 
            "region": "us-east-1", 
            "security_token": null, 
            "state": "present", 
            "validate_certs": true
        }
    }, 
    "name": {
        "response_metadata": {
            "http_status_code": 200, 
            "request_id": "6e4b8f69-9545-4534-a85d-6613a93ebcae"
        }
    }
}
```

**Comment by MichaelBaydoun** [comment]
I believe github allows the merger to squash commits at merge time now

**Comment by gregdek** [comment]
Since we can now squash commits at merge time, and since this module has received a fair amount of attention in our review meetings, I think it's time to give this one a shipit. Thanks @MichaelBaydoun :)

File: cloud/amazon/ec2_customer_gateway.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+DOCUMENTATION = '''
+---
+module: ec2_customer_gateway
+short_description: Manage an AWS customer gateway
+description:
+    - Manage an AWS customer gateway
+version_added: "2.2"
+author: Michael Baydoun (@MichaelBaydoun)
+requirements: [ botocore boto3 ]
```
Comment (abadger) [inline cloud/amazon/ec2_customer_gateway.py]: requirements is a list, So need to separate those with a comma.

File: cloud/amazon/ec2_customer_gateway.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+    def __init__(self, module):
+        self.module = module
+
+        try:
+            region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+            if not region:
+                module.fail_json(msg="Region must be specified as a parameter, in EC2_REGION or AWS_REGION environment variables or in boto configuration file")
+            self.ec2 = boto3_conn(module, conn_type='client', resource='ec2', region=region, endpoint=ec2_url, **aws_connect_kwargs)
+        except ClientError, e:
```
Comment (abadger) [inline cloud/amazon/ec2_customer_gateway.py:131]: ClientError is undefined here, is it botocore.exceptions.ClientError?  (Also need to import this).

**Comment by MichaelBaydoun** [comment]
ready_for_review. or merge, or something.

**Comment by gregdek** [comment]
Thanks @MichaelBaydoun for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by abadger** [comment]
Thanks @MichaelBaydoun for all your work on this!  Merged to devel.

**Pull request merged**